src/ExternalApi/Ada/Service/AdaProgrammeService.php

```
@@ -0,0 +1,143 @@
+            function ($results) use ($cacheItem, $limit) {
+                // $results contains the results of all promises in $promises indexed by the keys supplied in
+                // that array
+                return $this->parseAggregateResponses($results, $cacheItem, $limit);
+            },
+            // onRejected
+            function ($reason) {
+                // Logging should already have been handled by HttpApiClient. If any request fails
+                // return nothing and get on with our lives, data is not critical here
+                return [];
+            }
+        );
+        return $aggregatePromise;
+    }
+
+    private function parseAggregateResponses(array $results, CacheItemInterface $cacheItem, int $limit): array
+    {
+        $uniqueProgrammes = array_unique(array_merge($results['relatedByTag'], $results['relatedByBrand'], $results['relatedByCategory']), SORT_REGULAR);
+        $uniqueProgrammes = array_slice($uniqueProgrammes, 0, $limit);
+
+        if (empty($uniqueProgrammes)) {
+            $this->cache->setItem($cacheItem, [], CacheInterface::NORMAL);
+            return [];
+        }
+
+        $pids = [];
+        foreach ($uniqueProgrammes as $uniqueProgramme) {
+            $pids[] = new Pid($uniqueProgramme['pid']);
+        }
+        // Collect all the Programmes objects from the Programmes service using the pids array
+        $programmeItems = $this->programmesService->findByPids($pids);
+        $relatedProgrammes = [];
+        foreach ($uniqueProgrammes as $key => $item) {
+            $relatedProgrammes[] = $this->mapper->mapItem($programmeItems[$key], $item);
+        }
+
+        $this->cache->setItem($cacheItem, $relatedProgrammes, CacheInterface::MEDIUM);
+        return $relatedProgrammes;
+    }
+
+    private function requestRelatedProgrammeItems(string $pid, $scope = null, Pid $countContextPid = null, $threshold = null, $order = null, $orderDirection = null, $limit = null, $page = 1)
```

It seems this is always called with an order of 'random'.

---

Setting the default order to Random doesn't seem like a great idea to me, maybe Alphabetical by default if we have to have one?

---

Where in the code are we using anything other than 'random'?

---

Okay, fair point, we're not, it just seems very counter-intuitive to default to random. Like the default play format on music being shuffle play...